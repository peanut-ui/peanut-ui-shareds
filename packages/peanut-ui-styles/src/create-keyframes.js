import { keyframes } from '@emotion/css'

/**
 * Function for creating keyframes and generating them into unique animation classname.
 *
 * @param {Object} values - object css keyframes.
 * @returns {String} string css classname.
 */

export const createKeyframes = (values) => keyframes(values)
