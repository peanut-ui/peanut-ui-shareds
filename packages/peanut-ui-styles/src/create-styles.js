import { css, cx } from '@emotion/css'
import { isObject, isFunction } from '@chakra-ui/utils'
import { isTokenVariable, toCssVariable } from './utils/css-variable.util'
import { cssProperties } from './utils/css-property.util'

/**
 * @desc function for merging class names and adding a prefix.
 * @param {function} cx
 * @param {object} classes
 * @param {string} prefix
 * @return {object}
 */

const mergesCssClass = (cx, classes, prefix) => {
  return Object.keys(classes).reduce((previous, current) => {
    previous[current] = cx(classes[current], prefix ? `peanut-${prefix}-${current}` : null)

    return previous
  }, {})
}

/**
 * @desc function for converting an array of class names into objects.
 * @param {array} values
 * @return {object}
 */

const entriesCssObject = (values) => {
  const result = {}

  Object.keys(values).forEach((index) => {
    const [key, value] = values[index]

    result[key] = value
  })

  return result
}

/**
 * @desc function for creating styles and generating them into unique class names.
 * @param {object} values
 * @return {function}
 */

export const createStyles = (values) => {
  const getValues = isFunction(values) ? values : () => values

  /**
   * @desc function for accessing class names data.
   * @param {object} params
   * @param {string} prefix
   * @return {object}
   */

  const getStyles = (params, prefix) => {
    const values = getValues(params)
    const classes = entriesCssObject(
      Object.keys(values).map((key) => {
        const resolveStyles = (values) => {
          let computed = {}

          for (let [key, value] of Object.entries(values)) {
            if (key in cssProperties) {
              if (key in cssProperties) key = cssProperties[key]
              if (isObject(value)) {
                computed[key] = resolveStyles(value)
              } else {
                if (isTokenVariable(value)) {
                  const { reference } = toCssVariable(value, 'peanut')
                  computed[key] = reference
                } else computed[key] = value
              }
            } else delete computed[key]
          }

          return computed
        }

        const merged = cx(css(resolveStyles(values[key])))

        return [key, merged]
      })
    )

    return { classes: mergesCssClass(cx, classes, prefix), cx }
  }

  return getStyles
}
