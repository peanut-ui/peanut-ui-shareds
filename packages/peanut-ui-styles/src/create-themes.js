import { isObject } from '@chakra-ui/utils'
import { toCssVariable } from './utils/css-variable.util'

/**
 * @desc function for creating additional themes.
 * @param {object} values
 * @return {object}
 */

export const createThemes = (values) => {
  let variables = {}
  let maps = {}
  let references = []

  const resolveThemes = (values) => {
    for (let [key, value] of Object.entries(values)) {
      if (isObject(value)) {
        references.push(key)
        resolveThemes(value)
      } else {
        const keys = references.join('.').concat('.', key)
        const { variable, reference } = toCssVariable(keys, 'peanut')

        variables[variable] = value
        maps[keys] = { value, variable, reference }
      }
    }

    references = []
  }

  resolveThemes(values)

  return { variables, maps }
}
