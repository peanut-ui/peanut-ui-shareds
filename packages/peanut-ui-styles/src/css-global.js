import { injectGlobal } from '@emotion/css'

const toKebabCase = (value) => value.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`)

/**
 * @desc function for generating styles and injecting them into global scope.
 * @param {object} values
 * @return {string}
 */

export const cssGlobal = (values) => {
  let merged = ''

  for (let [key, value] of Object.entries(values)) {
    let computed = ''

    Object.keys(value).map((key) => (computed += `${toKebabCase(key)}:${value[key]}`.concat(';')))
    merged += `${key}{${computed}}`.replace(/\s+/g, '')
  }

  injectGlobal`${merged}`

  return merged
}
