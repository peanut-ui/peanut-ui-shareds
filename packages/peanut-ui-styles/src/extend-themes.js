/**
 * @desc function for merging or extending themes.
 * @param {object} values
 * @param {object} override
 * @return {object}
 */

export const extendThemes = (values, override) => {
  const variables = Object.assign({}, values?.variables, override?.variables)
  const maps = Object.assign({}, values?.maps, override?.maps)

  return { variables, maps }
}
