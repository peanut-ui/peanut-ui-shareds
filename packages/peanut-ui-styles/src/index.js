export * from './create-keyframes'
export * from './create-styles'
export * from './create-themes'
export * from './css-global'
