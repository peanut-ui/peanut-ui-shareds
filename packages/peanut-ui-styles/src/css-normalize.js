import { cssGlobal } from './css-global'
import { cssNormalizeStyles } from './utils/css-normalize.util'

export const cssNormalize = () => cssGlobal(cssNormalizeStyles)
