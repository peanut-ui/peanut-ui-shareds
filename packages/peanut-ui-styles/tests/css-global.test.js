import { cssGlobal } from '../src'

const styles = {
  '*': {
    margin: 0,
    padding: 0,
  },
}

describe('@peanut-ui/peanut-ui-styles/css-global', () => {
  test('should create or check match snapshot', () => {
    expect(cssGlobal(styles)).toMatchSnapshot()
  })
})
