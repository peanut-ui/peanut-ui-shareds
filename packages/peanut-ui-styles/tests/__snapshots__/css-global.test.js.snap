// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@peanut-ui/peanut-ui-styles/css-global should create or check match snapshot 1`] = `"*{margin:0;padding:0;}"`;
